Source/WebCore/animation/AnimationTimelinesController.cpp

```diff
@@ -589,20 +603,50 @@ void AnimationTimelinesController::setTimelineForName(const AtomString& name, co
         return !timeline->isInactiveStyleOriginatedTimeline();
     });
 
-    if (!hasNamedTimeline) {
+    // If we don't have an active named timeline yet and deferral is allowed,
+    // just register a pending timeline attachment operation so we can try again
+    // when style has resolved.
+    if (!hasNamedTimeline && allowsDeferral == AllowsDeferral::Yes) {
         m_pendingAttachOperations.append({ styleable, name, animation });
         return;
     }
 
-    auto& timelines = it->value;
-    if (RefPtr timeline = determineTimelineForElement(timelines, styleable, relatedTimelineScopeElements(name))) {
-        LOG_WITH_STREAM(Animations, stream << "AnimationTimelinesController::setTimelineForName: " << name << " styleable: " << styleable << " attaching to timeline of element: " << originatingElement(*timeline));
-        animation.setTimeline(WTFMove(timeline));
-        return;
-    }
-    animation.setTimeline(nullptr);
+    auto timelineScopeElements = relatedTimelineScopeElements(name);
 
-    m_pendingAttachOperations.append({ styleable, name, animation });
+    if (!hasNamedTimeline) {
+        // First, determine whether the name is within scope, ie. whether a parent element
+        // has a `timeline-scope` property that contains this timeline name.
+        auto nameIsWithinScope = [&] {
+            for (auto timelineScopeElement : timelineScopeElements) {
+                ASSERT(timelineScopeElement.element());
+                if (styleable == timelineScopeElement.styleable() || styleable.element.isDescendantOf(*timelineScopeElement.element()))
+                    return true;
+            }
+            return false;
+        }();
+
+        ASSERT(allowsDeferral == AllowsDeferral::No);
+        // We don't have an active named timeline and yet we must set a timeline since
+        // we've already dealt with the defferal case before. There are two cases:
+        //     1. the name is within scope and we should create a placeholder inactive
+        //        scroll timeline, or,
+        //     2. the name is not within scope and the timeline is null.
+        if (nameIsWithinScope)
+            animation.setTimeline(&inactiveNamedTimeline(name));
+        else {
+            animation.setTimeline(nullptr);
+            // Since we have no timelines defined for this name yet, we need
+            // to keep a pending operation such that we may attach the named
+            // timeline should it appear.
+            m_pendingAttachOperations.append({ styleable, name, animation });
+        }
+    } else {
+        auto& timelines = it->value;
+        if (RefPtr timeline = determineTimelineForElement(timelines, styleable, timelineScopeElements)) {
+            LOG_WITH_STREAM(Animations, stream << "AnimationTimelinesController::setTimelineForName: " << name << " styleable: " << styleable << " attaching to timeline of element: " << originatingElement(*timeline));
```

Comment:
Is this meant to be included?

Reply:
It was existing code that was moved here (see above in the red block).